classes/log_fetcher/course_log_fetcher.php

```diff
@@ -0,0 +1,235 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Course log fetcher
+ *
+ * @package   tool_s3logs
+ * @author    Kenneth Hendricks <kennethhendricks@catalyst-au.net>
+ * @copyright Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_s3logs\log_fetcher;
+
+use tool_s3logs\client\s3_client;
+
+class course_log_fetcher {
+
+    /**
+     * Fetches logs for multiple courses at once and writes to seperate files.
+     *
+     * @param integer $courseids Course ids.
+     * @param string  $logfolder Log folder path.
+     * @return void
+     */
+    public static function fetch_logs($courseids, $logfolder) {
+
+        echo "Setting up course log files ... \n";
+        $courseloghandles = self::get_course_log_file_handles($courseids, $logfolder);
+        self::write_headers_to_course_log_files($courseloghandles);
+
+        echo "Getting S3 log file keys ... \n";
+        $s3logkeys = self::get_s3_log_keys();
+
+        echo "Fetching logs from S3 \n";
+        self::write_s3_logs_to_course_logs($courseids, $courseloghandles, $s3logkeys);
+
+        echo "Fetching logs from DB \n";
+        self::write_current_database_logs_to_course_logs($courseids, $courseloghandles);
+
+        self::close_course_log_file_handles($courseloghandles);
+
+        echo "Finished pulling logs \n";
+    }
+
+    /**
+     * Pulls s3 logs, parses them and writes relevent course log entries to
+     * respective files.
+     *
+     * @param array $courseids        Course ids.
+     * @param array $courseloghandles Course log file handles.
+     * @param array $s3logkeys        S3 log keys
+     * @return void
+     */
+    private static function write_s3_logs_to_course_logs($courseids, $courseloghandles, $s3logkeys) {
+        $currentlogcount = 0;
+        $totallogs = count($s3logkeys);
+        foreach ($s3logkeys as $s3logkey) {
+            $currentlogcount++;
+
+            echo "\nDownloading $s3logkey ($currentlogcount/$totallogs) \n";
+            $loghandle = self::download_s3_log_file($s3logkey);
```

Comment:
do we want this loop to continue if we can't get one of the files? Or we happy for the whole thing to explode?

Reply:
whole thing to explode IMO